### PR TITLE
VReplication: Guard against unsafe _vt.vreplication writes

### PR DESCRIPTION
**Pull request by mattlord**
## Description

There is no limit on the number of concurrent active workflows you can have in a keyspace. This means that any time we're updating the `_vt.vreplication` table we should uniquely identify a workflow using the id (unique stream for a unique workflow) or workflow name. If we do not, then in various situations we can perform errant steps that can lead to workflow failure or even data corruption.

This PR fixes the few places where we were NOT already doing this (actually, this part was extracted to https://github.com/vitessio/vitess/pull/14826 for backports) — with the exception of the Resharder/StreamMigrator as this is an exception where we DO want to affect multiple workflows (as we move them from the active to inactive shards for a Reshard).

This PR also adds a guardrail for `UPDATE` and `DELETE` statements to help prevent Vitess users and developers from accidentally introducing related bugs in the future. For example:
```
❯ vtctlclient VReplicationExec zone1-101 "delete from _vt.vreplicati" 2>/dev/null

WARNING: VReplicationExec is deprecated and will be removed in a future release. Please use 'Workflow -- <keyspace.workflow> <action>' instead.

VReplicationExec Error: rpc error: code = Unknown desc = TabletManager.VReplicationExec on zone1-0000000101: invalid table name: vreplicati

❯ vtctlclient VReplicationExec zone1-101 "delete from _vt.vreplication where state != 'Running'" 2>/dev/null

WARNING: VReplicationExec is deprecated and will be removed in a future release. Please use 'Workflow -- <keyspace.workflow> <action>' instead.

VReplicationExec Error: rpc error: code = Unknown desc = TabletManager.VReplicationExec on zone1-0000000100: unsafe WHERE clause in delete without the /*vt+ ALLOW_UNSAFE_VREPLICATION_WRITE */ comment directive:  where state != 'Running'; should be using = or in with at least one of the following columns: id, workflow
```

## Related Issue(s)

  - Fixes: https://github.com/vitessio/vitess/issues/14795
  - Superset of: https://github.com/vitessio/vitess/pull/14826

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtctl/workflow/stream_migrator.go [hunk 1]
```diff
@@ -156,7 +155,8 @@ func (sm *StreamMigrator) CancelMigration(ctx context.Context) {
 	_ = sm.deleteTargetStreams(ctx)
 
 	err := sm.ts.ForAllSources(func(source *MigrationSource) error {
-		query := fmt.Sprintf("update _vt.vreplication set state='Running', stop_pos=null, message='' where db_name=%s and workflow != %s", encodeString(source.GetPrimary().DbName()), encodeString(sm.ts.ReverseWorkflowName()))
+		query := fmt.Sprintf("update _vt.vreplication set state='Running', stop_pos=null, message='' where db_name=%s and workflow=%s",
+			encodeString(source.GetPrimary().DbName()), encodeString(sm.ts.WorkflowName()))
 		_, err := sm.ts.VReplicationExec(ctx, source.GetPrimary().Alias, query)
 		return err
```
Comment (harshit-gangal) [inline go/vt/vtctl/workflow/stream_migrator.go:161]: can you explain the difference here.
Looks like earlier is used to bulk update as `workflow !=` would match more rows than the new change `workflow =`. I am not sure the implications here but just stating it out.

Comment (mattlord) [reply go/vt/vtctl/workflow/stream_migrator.go:161]: In general the difference is what is described in the PR. 🙂 In this specific case, it's taking an unsafe query which was making an assumption about the data — namely that there's only one workflow in the keyspace and thus only records for the forward and reverse variants of it — and making it a safe query which only updates records for the specific workflow (and forward/reverse variant of it) that it should be.

File: go/vt/vttablet/tabletmanager/rpc_server.go [hunk 2]
```diff
@@ -70,7 +70,7 @@ func (tm *TabletManager) HandleRPCPanic(ctx context.Context, name string, args,
 	if *err != nil {
 		// error case
 		log.Warningf("TabletManager.%v(%v)(on %v from %v) error: %v", name, args, topoproto.TabletAliasString(tm.tabletAlias), from, (*err).Error())
-		*err = vterrors.Wrapf(*err, "TabletManager.%v on %v error: %v", name, topoproto.TabletAliasString(tm.tabletAlias), (*err).Error())
+		*err = vterrors.Wrapf(*err, "TabletManager.%v on %v", name, topoproto.TabletAliasString(tm.tabletAlias))
```
Comment (mattlord) [inline go/vt/vttablet/tabletmanager/rpc_server.go:73]: This removes an improper usage of wrapping which led to the error/cause being duplicated.

File: go/test/endtoend/vreplication/cluster_test.go [hunk 3]
```diff
@@ -385,8 +385,19 @@ func (vc *VitessCluster) CleanupDataroot(t *testing.T, recreate bool) {
...
+	var err error
+	retries := 3
+	for i := 1; i <= retries; i++ {
+		if err = os.RemoveAll(dir); err == nil {
+			log.Infof("Deleted vtdataroot %q", dir)
+			break
+		}
+		log.Errorf("Failed to delete vtdataroot (attempt %d of %d) %q: %v", i, retries, dir, err)
+		time.Sleep(1 * time.Second)
+	}
```
Comment (mattlord) [inline go/test/endtoend/vreplication/cluster_test.go:400]: This is something unrelated that I happened to see during all of the workflow runs in this PR.

File: go/vt/vttablet/tabletmanager/vreplication/controller_plan.go [hunk 4]
```diff
@@ -50,6 +51,69 @@ const (
...
+				node.Operator == sqlparser.InOp {
+				selective = true  // This is a safe statement
+				return false, nil // We can stop walking
+			}
+		default:
+		}
+		return true, nil
+	}, where)
+	return selective
+}
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletmanager/vreplication/controller_plan.go:97]: Nice one!

Comment (shlomi-noach) [reply go/vt/vttablet/tabletmanager/vreplication/controller_plan.go:97]: Is this only limited to `controller_plan` or should this be expanded to include other places?

**Review by shlomi-noach — APPROVED** [review]
It's pretty wild, good catch.

File: go/vt/vttablet/tabletmanager/vreplication/controller_plan.go (hunk 4 above)
Comment (mattlord) [reply go/vt/vttablet/tabletmanager/vreplication/controller_plan.go:97]: I feel like it's probably a good thing to expand over time, but at this point vreplication.Engine.Exec() is the critical one and what's done here.

File: go/vt/vtctl/workflow/stream_migrator.go [hunk 5]
```diff
@@ -201,8 +201,8 @@ func (sm *StreamMigrator) StopStreams(ctx context.Context) ([]string, error) {
 /* tablet streams */
 
 func (sm *StreamMigrator) readTabletStreams(ctx context.Context, ti *topo.TabletInfo, constraint string) ([]*VReplicationStream, error) {
-	query := fmt.Sprintf("select id, workflow, source, pos, workflow_type, workflow_sub_type, defer_secondary_keys from _vt.vreplication where db_name=%s and workflow != %s",
-		encodeString(ti.DbName()), encodeString(sm.ts.ReverseWorkflowName()))
+	query := fmt.Sprintf("select id, workflow, source, pos, workflow_type, workflow_sub_type, defer_secondary_keys from _vt.vreplication where db_name=%s and workflow=%s",
```
Comment (rohit-nayak-ps) [inline go/vt/vtctl/workflow/stream_migrator.go:204]: We only migrate streams for reshard workflows. The above code is expected to select all appropriate streams (Materialize/CreateLookupVindex workflow types) that need to be moved to the new shards.

**Review by rohit-nayak-ps — CHANGES_REQUESTED** [review]
The assumptions related to stream migrations are incorrect here, as commented around those changes. 

The fix for reverse streams in traffic_switcher is good. 

The change for resharder makes the code clearer , but it is a noop, since there are no other streams on the target at that time.

File: go/test/endtoend/vreplication/cluster_test.go [hunk 6]
```diff
@@ -385,8 +385,19 @@ func (vc *VitessCluster) CleanupDataroot(t *testing.T, recreate bool) {
 		return
 	}
 	dir := vc.ClusterConfig.vtdataroot
-	log.Infof("Deleting vtdataroot %s", dir)
-	err := os.RemoveAll(dir)
+	// The directory cleanup sometimes fails with a "directory not empty" error as
+	// everything in the test is shutting down and cleaning up. So we retry a few
+	// times to deal with that non-problematic and ephemeral issue.
+	var err error
+	retries := 3
```
[the remark this answers is not in the document]
Comment (mattlord) [reply go/test/endtoend/vreplication/cluster_test.go:392]: It's only been seen in the CI AFAIK, and so far 3 retries with a 1 second delay seems to have been enough (we've done essentially the same thing in several places now).

**Review by rohit-nayak-ps — COMMENTED** [review]
A couple of comments, otherwise looks great.

File: go/vt/vtctl/workflow/switcher_interface.go [hunk 7]
```diff
@@ -25,7 +25,7 @@ import (
 
 type iswitcher interface {
 	lockKeyspace(ctx context.Context, keyspace, action string) (context.Context, func(*error), error)
-	cancelMigration(ctx context.Context, sm *StreamMigrator)
+	cancelStreamMigrations(ctx context.Context, sm *StreamMigrator)
```
[the remark this answers is not in the document]
Comment (mattlord) [reply go/vt/vtctl/workflow/switcher_interface.go:28]: Good point. Initially I was just going to rename the one that has a StreamMigrator receiver and got carried away... 😄 I'll change the ones using the switcher interface back as those are more generic.

File: go/vt/wrangler/vexec.go [hunk 8]
```diff
@@ -478,9 +478,9 @@ func (wr *Wrangler) execWorkflowAction(ctx context.Context, workflow, keyspace,
 	return wr.runVexec(ctx, workflow, keyspace, query, callback, dryRun)
 }
 
-// WorkflowTagAction sets or clears the tags for a workflow in a keyspace
+// WorkflowTagAction sets or clears the tags for a workflow in a keyspace.
 func (wr *Wrangler) WorkflowTagAction(ctx context.Context, keyspace string, workflow string, tags string) (map[*topo.TabletInfo]*sqltypes.Result, error) {
-	query := fmt.Sprintf("update _vt.vreplication set tags = %s", encodeString(tags))
+	query := fmt.Sprintf("update _vt.vreplication set tags = %s where workflow = %s", encodeString(tags), encodeString(workflow))
```
[the remark this answers is not in the document]
Comment (mattlord) [reply go/vt/wrangler/vexec.go:483]: Yeah, I see it adds the workflow name if it's not in a where clause already in `,vexec.addDefaultWheres()`

**Review by rohit-nayak-ps — APPROVED** [review]

**Pull request merged**